File: spring/demo/src/main/java/com/example/demo/config/AppConfig.java
package com.example.demo.config;
import java.util.HashMap;
import java.util.Map;

public class AppConfig {
    private static AppConfig instance;
    private final Map<String, Object> configs = new HashMap<>();

    private AppConfig() {
        configs.put("dbHost", "localhost");
        configs.put("dbPort", 5432);
        configs.put("shopName", "Highland Coffee");
        configs.put("maxOrders", 100);
    }

    public static synchronized AppConfig getInstance() {
        if (instance == null) instance = new AppConfig();
        return instance;
    }

    public Object getConfig(String key) { return configs.get(key); }
    public void setConfig(String key, Object value) { configs.put(key, value); }
}

File: nodejs/src/config/AppConfig.ts
// 1. Define an interface for the configuration object
interface ConfigMap {
    dbHost: string;
    dbPort: number;
    shopName: string;
    maxOrders: number;
}

class AppConfig {
    private static instance: AppConfig;
    private configs: ConfigMap;

    // 2. The constructor MUST be private in a Singleton
    private constructor() {
        this.configs = {
            dbHost: "localhost",
            dbPort: 5432,
            shopName: "Highland Coffee",
            maxOrders: 100,
        };
    }

    // 3. Proper instantiation logic
    public static getInstance(): AppConfig {
        if (!AppConfig.instance) {
            AppConfig.instance = new AppConfig();
        }
        return AppConfig.instance;
    }

    // 4. Use Generics and 'keyof' for strict type safety
    public getConfig<K extends keyof ConfigMap>(key: K): ConfigMap[K] {
        return this.configs[key];
    }

    public setConfig<K extends keyof ConfigMap>(
        key: K,
        value: ConfigMap[K],
    ): void {
        this.configs[key] = value;
    }
}

export default AppConfig;
